File: backend/app/routers/registries.py
"""Per-registry credential store backed by a JSON file (mode 0600).

Credentials are used to authenticate image pulls against private registries.
The stored format is intentionally simple so it can be inspected and edited
out-of-band by operators if needed:

    {"registries": {"<name>": {"url": "...", "username": "...",
                               "password": "...", "email": "..."}}}
"""
from __future__ import annotations

import json
import os
import threading

from docker.errors import APIError
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from ..auth import User, authenticate, require_admin
from ..config import settings
from ..docker_client import get_client

router = APIRouter(prefix="/api/registries", tags=["registries"])

_LOCK = threading.Lock()


class RegistryIn(BaseModel):
    name: str = Field(..., description="Local nickname for this credential set")
    url: str = Field(
        "https://index.docker.io/v1/",
        description="Registry URL (e.g. https://ghcr.io, https://my.registry.example/v2/)",
    )
    username: str
    password: str
    email: str | None = None


class RegistryPublic(BaseModel):
    name: str
    url: str
    username: str
    email: str | None = None


def _load() -> dict:
    p = settings.registries_file
    try:
        with open(p, "r", encoding="utf-8") as fh:
            data = json.load(fh)
            return data.get("registries", {}) if isinstance(data, dict) else {}
    except FileNotFoundError:
        return {}
    except (OSError, json.JSONDecodeError):
        return {}


def _save(creds: dict) -> None:
    p = settings.registries_file
    parent = os.path.dirname(p) or "."
    try:
        os.makedirs(parent, exist_ok=True)
    except OSError as exc:
        raise HTTPException(
            status_code=500,
            detail=f"Cannot create directory for registries file: {exc}",
        ) from exc
    tmp = f"{p}.tmp"
    with open(tmp, "w", encoding="utf-8") as fh:
        json.dump({"registries": creds}, fh, indent=2)
    try:
        os.chmod(tmp, 0o600)
    except OSError:
        pass
    os.replace(tmp, p)


def get_registry_auth(name: str) -> dict | None:
    """Public helper used by the images router when pulling."""
    if not name:
        return None
    creds = _load()
    rec = creds.get(name)
    if not rec:
        return None
    out: dict = {"username": rec["username"], "password": rec["password"]}
    if rec.get("email"):
        out["email"] = rec["email"]
    if rec.get("url"):
        out["serveraddress"] = rec["url"]
    return out


@router.get("", response_model=list[RegistryPublic])
def list_registries(_: User = Depends(authenticate)) -> list[RegistryPublic]:
    with _LOCK:
        creds = _load()
    return [
        RegistryPublic(
            name=name,
            url=rec.get("url", ""),
            username=rec.get("username", ""),
            email=rec.get("email"),
        )
        for name, rec in sorted(creds.items())
    ]


@router.put("/{name}")
@router.post("")
def upsert(req: RegistryIn, name: str | None = None, _: User = Depends(require_admin)) -> dict:
    target = name or req.name
    if not target:
        raise HTTPException(status_code=400, detail="Registry name required")
    with _LOCK:
        creds = _load()
        creds[target] = {
            "url": req.url,
            "username": req.username,
            "password": req.password,
            "email": req.email,
        }
        _save(creds)
    return {"name": target, "saved": True}


@router.delete("/{name}")
def delete(name: str, _: User = Depends(require_admin)) -> dict:
    with _LOCK:
        creds = _load()
        if name not in creds:
            raise HTTPException(status_code=404, detail="Registry not found")
        del creds[name]
        _save(creds)
    return {"removed": name}


@router.post("/{name}/test")
def test_login(name: str, _: User = Depends(require_admin)) -> dict:
    with _LOCK:
        creds = _load()
    if name not in creds:
        raise HTTPException(status_code=404, detail="Registry not found")
    rec = creds[name]
    try:
        result = get_client().login(
            username=rec["username"],
            password=rec["password"],
            email=rec.get("email"),
            registry=rec.get("url") or None,
            reauth=True,
        )
        return {"ok": True, "result": result}
    except APIError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

File: backend-node/src/routes/registries.js
// Per-registry credential store backed by a JSON file (mode 0600).
import fs from 'node:fs/promises';
import path from 'node:path';
import { Type } from '@sinclair/typebox';
import { settings } from '../config.js';
import { getClient } from '../docker-client.js';
import { asyncHandler, HttpError } from '../util.js';
import { createApiRouter } from '../route-builder.js';
import {
  PassThroughObject,
  RegistryPublic,
  RegistryRequest,
  RegistryUpdateRequest,
} from '../schemas/index.js';

const r = createApiRouter('/api/registries', { tag: 'registries' });

let writeLock = Promise.resolve();
function withLock(fn) {
  const next = writeLock.then(fn, fn);
  writeLock = next.catch(() => {});
  return next;
}

async function load() {
  try {
    const raw = await fs.readFile(settings.registriesFile, 'utf8');
    const data = JSON.parse(raw);
    return (data && data.registries) || {};
  } catch (err) {
    return {};
  }
}
async function save(creds) {
  const file = settings.registriesFile;
  const dir = path.dirname(file);
  try { await fs.mkdir(dir, { recursive: true }); }
  catch (err) { throw new HttpError(500, `Cannot create directory for registries file: ${err.message}`); }
  const tmp = `${file}.tmp`;
  await fs.writeFile(tmp, JSON.stringify({ registries: creds }, null, 2));
  try { await fs.chmod(tmp, 0o600); } catch {}
  await fs.rename(tmp, file);
}

export async function getRegistryAuth(name) {
  if (!name) return null;
  const creds = await load();
  const rec = creds[name];
  if (!rec) return null;
  const out = { username: rec.username, password: rec.password };
  if (rec.email) out.email = rec.email;
  if (rec.url) out.serveraddress = rec.url;
  return out;
}

const NameParam = Type.Object({ name: Type.String() }, { additionalProperties: false });

r.get(
  '/',
  {
    summary: 'List stored registry credentials',
    responses: { 200: Type.Array(RegistryPublic) },
  },
  asyncHandler(async (_req, res) => {
    const creds = await load();
    const out = Object.entries(creds)
      .sort(([a], [b]) => a.localeCompare(b))
      .map(([name, rec]) => ({ name, url: rec.url || '', username: rec.username || '', email: rec.email || null }));
    res.json(out);
  }),
);

r.post(
  '/',
  {
    summary: 'Add or replace a registry credential set',
    admin: true,
    destructive: true,
    body: RegistryRequest,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    const b = req.body;
    await withLock(async () => {
      const creds = await load();
      creds[b.name] = { url: b.url || 'https://index.docker.io/v1/', username: b.username, password: b.password, email: b.email || null };
      await save(creds);
    });
    res.json({ name: b.name, saved: true });
  }),
);

r.put(
  '/:name',
  {
    summary: 'Add or replace by path (name comes from URL)',
    admin: true,
    destructive: true,
    params: NameParam,
    body: RegistryUpdateRequest,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    const name = req.params.name;
    const b = { ...req.body, name };
    if (!name) throw new HttpError(400, 'name is required');
    if (!b.username || !b.password) throw new HttpError(400, 'username and password are required');
    await withLock(async () => {
      const creds = await load();
      creds[name] = { url: b.url || 'https://index.docker.io/v1/', username: b.username, password: b.password, email: b.email || null };
      await save(creds);
    });
    res.json({ name, saved: true });
  }),
);

r.delete(
  '/:name',
  {
    summary: 'Delete a registry credential set',
    admin: true,
    destructive: true,
    params: NameParam,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    await withLock(async () => {
      const creds = await load();
      if (!creds[req.params.name]) throw new HttpError(404, 'Registry not found');
      delete creds[req.params.name];
      await save(creds);
    });
    res.json({ removed: req.params.name });
  }),
);

r.post(
  '/:name/test',
  {
    summary: 'Test login against the saved credentials',
    admin: true,
    destructive: true,
    params: NameParam,
    responses: { 200: PassThroughObject },
  },
  asyncHandler(async (req, res) => {
    const creds = await load();
    const rec = creds[req.params.name];
    if (!rec) throw new HttpError(404, 'Registry not found');
    const docker = getClient();
    const result = await new Promise((resolve, reject) => {
      docker.modem.dial(
        {
          path: '/auth', method: 'POST',
          options: { username: rec.username, password: rec.password, email: rec.email || undefined, serveraddress: rec.url || undefined },
          statusCodes: { 200: true, 500: 'server error' },
        },
        (err, data) => (err ? reject(err) : resolve(data)),
      );
    });
    res.json({ ok: true, result });
  }),
);

export default r;
